packages/read-model/tests/Unit/Product/Listed/ListedProductViewFactoryTest.php

```diff
@@ -0,0 +1,156 @@
+<?php
+
+namespace Tests\ReadModelBundle\Unit\Product\Listed;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use PHPUnit\Framework\TestCase;
+use Shopsys\FrameworkBundle\Component\Domain\Domain;
+use Shopsys\FrameworkBundle\Component\Money\Money;
+use Shopsys\FrameworkBundle\Model\Pricing\Price;
+use Shopsys\FrameworkBundle\Model\Product\Availability\Availability;
+use Shopsys\FrameworkBundle\Model\Product\Flag\Flag;
+use Shopsys\FrameworkBundle\Model\Product\Pricing\ProductPrice;
+use Shopsys\FrameworkBundle\Model\Product\Product;
+use Shopsys\FrameworkBundle\Model\Product\ProductCachedAttributesFacade;
+use Shopsys\ReadModelBundle\Image\ImageView;
+use Shopsys\ReadModelBundle\Product\Action\ProductActionView;
+use Shopsys\ReadModelBundle\Product\Listed\ListedProductView;
+use Shopsys\ReadModelBundle\Product\Listed\ListedProductViewFactory;
+
+class ListedProductViewFactoryTest extends TestCase
+{
+    /**
+     * @dataProvider getProductsData
+     * @param int $id
+     * @param string $productName
+     * @param string $shortDescription
+     * @param string $availabilityName
+     * @param int $priceAmount
+     * @param \Shopsys\ReadModelBundle\Image\ImageView $imageView
+     * @param \Shopsys\ReadModelBundle\Product\Action\ProductActionView $productActionView
+     * @param array|\Doctrine\Common\Collections\ArrayCollection $flags
+     * @param int[] $expectedFlags
+     */
+    public function testCreateFromProduct(int $id, string $productName, string $shortDescription, string $availabilityName, int $priceAmount, ImageView $imageView, ProductActionView $productActionView, $flags, $expectedFlags): void
+    {
+        $domainMock = $this->createDomainMock();
+        $productCachedAttributesFacadeMock = $this->createProductCachedAttributesFacadeMock($priceAmount);
+
+        $listedProductViewFactory = new ListedProductViewFactory($domainMock, $productCachedAttributesFacadeMock);
+
+        $productMock = $this->createProductMock($id, $productName, $shortDescription, $availabilityName, $flags);
+
+        /** @var \Shopsys\ReadModelBundle\Product\Listed\ListedProductView $listedProductView */
+        $listedProductView = $listedProductViewFactory->createFromProduct($productMock, $imageView, $productActionView);
+
+        $expected = new ListedProductView($id, $productName, $shortDescription, $availabilityName, $this->createProductPrice($priceAmount), $expectedFlags, $productActionView, $imageView);
+
+        $this->assertEquals($expected, $listedProductView);
+    }
+
+    /**
+     * @param int $id
+     * @param string $name
+     * @param string $shortDescription
+     * @param string $availabilityName
+     * @param array|\Doctrine\Common\Collections\ArrayCollection $flags
```

Comment:
```suggestion
     * @param Flag[] $flags
```

Reply:
I changed it to \Shopsys\FrameworkBundle\Model\Product\Flag\Flag[]|\Doctrine\Common\Collections\ArrayCollection because it's important to me to test ArrayCollection behavior and like this, it explains why there is no typehint for flags in method argument.